File: CSharp/SpecStudioExamplesTestsCSharp/common/FrameValuesTyped.cs
namespace SpecStudioExamplesTestsCSharp.common
{
using System;
using System.Buffers;
using System.Collections.Generic;
using System.Globalization;
using System.Text;
using System.Text.Json;
using production;

    public class FrameValuesTyped
    {
        public int frame;
        public Pins roll1;
        public Pins roll2;
        public Pins roll3;
        public Score score;
        public Score totalScore;

        public FrameValuesTyped(int frame, Pins roll1, Pins roll2, Pins roll3, Score score, Score totalScore)
        {
            this.frame = frame;
            this.roll1 = roll1;
            this.roll2 = roll2;
            this.roll3 = roll3;
            this.score = score;
            this.totalScore = totalScore;
        }

        public FrameValuesString ToFrameValuesString()
        {
            return new FrameValuesString(
                Json.ToText(this.frame),
                Json.ToText(this.roll1),
                Json.ToText(this.roll2),
                Json.ToText(this.roll3),
                Json.ToText(this.score),
                Json.ToText(this.totalScore)
            );
        }

        public static List<FrameValuesString> ToStringList(List<FrameValuesTyped> list)
        {
            var result = new List<FrameValuesString>();
            foreach (var t in list) result.Add(t.ToFrameValuesString());
            return result;
        }

        public static List<FrameValuesTyped> FromStringList(List<FrameValuesString> list)
        {
            var result = new List<FrameValuesTyped>();
            foreach (var s in list) result.Add(s.ToFrameValuesTyped());
            return result;
        }

        public void WriteJson(Utf8JsonWriter w)
        {
            w.WriteStartObject();
            w.WriteNumber("frame", this.frame);
            w.WriteString("roll1", Json.ToText(this.roll1));
            w.WriteString("roll2", Json.ToText(this.roll2));
            w.WriteString("roll3", Json.ToText(this.roll3));
            w.WriteString("score", Json.ToText(this.score));
            w.WriteString("totalScore", Json.ToText(this.totalScore));
            w.WriteEndObject();
        }

        public string ToJSON()
        {
            var buffer = new ArrayBufferWriter<byte>();
            using (var w = new Utf8JsonWriter(buffer)) { WriteJson(w); }
            return Encoding.UTF8.GetString(buffer.WrittenSpan);
        }

        public static FrameValuesTyped FromJsonElement(JsonElement m)
        {
            return new FrameValuesTyped(
                Json.AsInt(Json.Require(m, "frame"), "frame"),
                new Pins(Json.AsString(Json.Require(m, "roll1"), "roll1")),
                new Pins(Json.AsString(Json.Require(m, "roll2"), "roll2")),
                new Pins(Json.AsString(Json.Require(m, "roll3"), "roll3")),
                new Score(Json.AsString(Json.Require(m, "score"), "score")),
                new Score(Json.AsString(Json.Require(m, "totalScore"), "totalScore"))
            );
        }

        public static FrameValuesTyped FromJSON(string json)
        {
            return FromJsonElement(Json.Parse(json));
        }

        public static string ToJSONList(List<FrameValuesTyped> list)
        {
            var buffer = new ArrayBufferWriter<byte>();
            using (var w = new Utf8JsonWriter(buffer))
            {
                w.WriteStartArray();
                foreach (var item in list) item.WriteJson(w);
                w.WriteEndArray();
            }
            return Encoding.UTF8.GetString(buffer.WrittenSpan);
        }

        public static List<FrameValuesTyped> FromJSONList(string json)
        {
            var result = new List<FrameValuesTyped>();
            var root = Json.Parse(json);
            Json.RequireArray(root, "FrameValuesTyped");
            foreach (var e in root.EnumerateArray()) result.Add(FromJsonElement(e));
            return result;
        }

        public override string ToString()
        {
            return $"Frame={frame}, Roll1={roll1}, Roll2={roll2}, Roll3={roll3}, Score={score}, TotalScore={totalScore}";
        }

        public override bool Equals(object? obj)
        {
            if (obj is not FrameValuesTyped other) return false;
            return object.Equals(this.frame, other.frame)
                && object.Equals(this.roll1, other.roll1)
                && object.Equals(this.roll2, other.roll2)
                && object.Equals(this.roll3, other.roll3)
                && object.Equals(this.score, other.score)
                && object.Equals(this.totalScore, other.totalScore);
        }

        public override int GetHashCode()
        {
            var h = new System.HashCode();
            h.Add(this.frame);
            h.Add(this.roll1);
            h.Add(this.roll2);
            h.Add(this.roll3);
            h.Add(this.score);
            h.Add(this.totalScore);
            return h.ToHashCode();
        }
    }
}

File: Java/SpectableJavaExamplesTest/src/test/java/spectable/common/FrameValuesTyped.java
package spectable.common;

import java.util.ArrayList;
import java.util.List;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.Objects;
import production.*;
import records.*;
import calculator.*;

public class FrameValuesTyped {
    public int frame;
    public Pins roll1;
    public Pins roll2;
    public Pins roll3;
    public Score score;
    public Score totalScore;

    public FrameValuesTyped(int frame, Pins roll1, Pins roll2, Pins roll3, Score score, Score totalScore) {
        this.frame = frame;
        this.roll1 = roll1;
        this.roll2 = roll2;
        this.roll3 = roll3;
        this.score = score;
        this.totalScore = totalScore;
    }

    public FrameValuesTyped(FrameValuesString s) {
        this.frame = Integer.parseInt(s.frame);
        this.roll1 = new Pins(s.roll1);
        this.roll2 = new Pins(s.roll2);
        this.roll3 = new Pins(s.roll3);
        this.score = new Score(s.score);
        this.totalScore = new Score(s.totalScore);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof FrameValuesTyped)) return false;
        FrameValuesTyped that = (FrameValuesTyped) o;
        return Objects.equals(frame, that.frame)
            && Objects.equals(roll1, that.roll1)
            && Objects.equals(roll2, that.roll2)
            && Objects.equals(roll3, that.roll3)
            && Objects.equals(score, that.score)
            && Objects.equals(totalScore, that.totalScore);
    }

    @Override
    public int hashCode() {
        return Objects.hash(frame, roll1, roll2, roll3, score, totalScore);
    }

    public FrameValuesString toFrameValuesString() {
        return new FrameValuesString(String.valueOf(frame), String.valueOf(roll1), String.valueOf(roll2), String.valueOf(roll3), String.valueOf(score), String.valueOf(totalScore));
    }

    public static List<FrameValuesTyped> fromStringList(List<FrameValuesString> list) {
        List<FrameValuesTyped> result = new ArrayList<>();
        for (FrameValuesString s : list) result.add(new FrameValuesTyped(s));
        return result;
    }

    public static List<FrameValuesString> toStringList(List<FrameValuesTyped> list) {
        List<FrameValuesString> result = new ArrayList<>();
        for (FrameValuesTyped t : list) result.add(t.toFrameValuesString());
        return result;
    }

    /** Internal plumbing for {@link Json}; use toJSON() for JSON text. */
    public Map<String, Object> toJsonValue() {
        Map<String, Object> m = new LinkedHashMap<String, Object>();
        m.put("frame", frame);
        m.put("roll1", roll1 == null ? null : roll1.toString());
        m.put("roll2", roll2 == null ? null : roll2.toString());
        m.put("roll3", roll3 == null ? null : roll3.toString());
        m.put("score", score == null ? null : score.toString());
        m.put("totalScore", totalScore == null ? null : totalScore.toString());
        return m;
    }

    public String toJSON() { return Json.write(toJsonValue()); }

    /** Internal plumbing for {@link Json}; use fromJSON(String) for JSON text. */
    public static FrameValuesTyped fromJsonValue(Map<String, Object> m) {
        if (m == null) return null;
        return new FrameValuesTyped(
            Json.asInt(Json.require(m, "frame"), "frame"),
            new Pins(Json.asString(Json.require(m, "roll1"), "roll1")),
            new Pins(Json.asString(Json.require(m, "roll2"), "roll2")),
            new Pins(Json.asString(Json.require(m, "roll3"), "roll3")),
            new Score(Json.asString(Json.require(m, "score"), "score")),
            new Score(Json.asString(Json.require(m, "totalScore"), "totalScore")));
    }

    public static FrameValuesTyped fromJSON(String json) {
        return fromJsonValue(Json.parseObject(json));
    }

    public static String toJSONList(List<FrameValuesTyped> list) {
        List<Object> arr = new ArrayList<Object>();
        if (list != null)
            for (FrameValuesTyped item : list) arr.add(item == null ? null : item.toJsonValue());
        return Json.write(arr);
    }

    public static List<FrameValuesTyped> fromJSONList(String json) {
        List<FrameValuesTyped> result = new ArrayList<FrameValuesTyped>();
        for (Object e : Json.parseArray(json))
            result.add(fromJsonValue(Json.asObject(e, "FrameValuesTyped")));
        return result;
    }
}
